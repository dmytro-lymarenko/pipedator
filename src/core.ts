export interface ValidationError {
	message: string;
	path: string[]; // ['a', '0', 'test'] which is equivalent to 'a[0].test' of the value
}

export interface ValidationContext {
	value: any;
	path: string[];
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	pipe: (...validators: Validator[]) => Validator;
}

export function pipe(...validators: Validator[]) {
	return createValidator({
		validate: (value, ctx) => {
			// find the first error in pipe
			let i: number = 0;
			let error: ValidationError | null = null;

			while (i < validators.length && error === null) {
				error = validators[i].validate(value, ctx);
				i = i + 1;
			}

			return error;
		},
	});
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator(options: CreateValidatorOptions): Validator {
	function validate(value: any, ctx?: ValidationContext) {
		return options.validate(value, ctx || { value, path: [] });
	}

	return {
		validate,
		pipe: (...validators) =>
			pipe(
				createValidator({ validate }),
				...validators
			),
	};
}
