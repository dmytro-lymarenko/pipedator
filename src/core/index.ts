export interface ValidationError {
	message: string;
	path: string;
	children: ValidationError[] | null;
}

export interface ValidationContext {
	value: any;
	path: string[];
	rootValue: any;
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator<ValidValue = any> {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	isValid: (value: any) => value is ValidValue;
}

export function findFirstError(
	getValidator: (index: number) => Validator,
	getValue: (index: number) => any,
	getContext: (index: number) => ValidationContext,
	length: number
): { error: ValidationError | null; index: number } {
	let i: number = 0;
	let error: ValidationError | null = null;

	while (i < length && error === null) {
		error = getValidator(i).validate(getValue(i), getContext(i));
		i = i + 1;
	}

	return { error, index: i - 1 };
}

export function getFirstErrors(
	getValidator: (index: number) => Validator,
	getValue: (index: number) => any,
	getContext: (index: number) => ValidationContext,
	length: number
): ValidationError[] {
	const errors: ValidationError[] = [];
	let i: number = 0;
	let error: ValidationError | null = null;

	do {
		error = getValidator(i).validate(getValue(i), getContext(i));
		if (error) {
			errors.push(error);
		}
		i = i + 1;
	} while (i < length && error !== null);

	return errors;
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator<ValidValue = any>(options: CreateValidatorOptions): Validator<ValidValue> {
	function validate(value: any, ctx?: ValidationContext) {
		const context: ValidationContext = ctx || {
			value,
			path: [],
			rootValue: value,
		};

		return options.validate(value, context);
	}

	function isValid(value: any): value is ValidValue {
		return validate(value) === null;
	}

	return {
		validate,
		isValid,
	};
}

export function validate(validator: Validator, value: any) {
	const ctx: ValidationContext = {
		value,
		path: [],
		rootValue: value,
	};

	return validator.validate(value, ctx);
}

export function getCurrentPath(ctx: ValidationContext): string {
	return ctx.path.length === 0 ? '' : ctx.path[ctx.path.length - 1];
}

export interface ValidationRef {
	(ctx: ValidationContext): any;
}

export function ref(path: string[]): ValidationRef {
	return (ctx: ValidationContext) => path.reduce((v, key) => v && v[key], ctx.rootValue);
}
