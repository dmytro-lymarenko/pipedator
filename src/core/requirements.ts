export interface BaseRequirement {
	value: any;
	path: string[];
	message: string;
}

export interface RequirementFactory<R extends Requirement = Requirement> {
	(path: string[], value: any): R;
}

export type Requirement = SingleRequirement | DependenceRequirement | AlternativeRequirement | ConcatenationRequirement;

export interface SingleRequirement extends BaseRequirement {
	type: 'single';
}

export interface DependenceRequirement extends BaseRequirement {
	type: 'dependence';
	requirement: Requirement;
}

export interface AlternativeRequirement extends BaseRequirement {
	type: 'alternative';
	requirements: Requirement[];
}

export interface ConcatenationRequirement extends BaseRequirement {
	type: 'concatenation';
	requirements: Requirement[];
}

export function isRequirement(requirement: Requirement | null): requirement is Requirement {
	return requirement !== null;
}

// Factories

export function singleRequirementFactory(message: string): RequirementFactory<SingleRequirement> {
	return (path, value) => ({
		path,
		value,
		message,
		type: 'single',
	});
}

export function dependenceRequirementFactory(
	message: string,
	requirement: Requirement
): RequirementFactory<DependenceRequirement> {
	return (path, value) => ({
		path,
		value,
		requirement,
		message,
		type: 'dependence',
	});
}

export function alternativeRequirementFactory(
	message: string,
	requirements: Requirement[]
): RequirementFactory<AlternativeRequirement | DependenceRequirement> {
	if (requirements.length < 1) {
		throw new Error('There must be 1 or more requirements');
	}

	if (requirements.length === 1) {
		return dependenceRequirementFactory(message, requirements[0]);
	}

	return (path, value) => ({
		path,
		value,
		message,
		type: 'alternative',
		requirements: requirements.reduce(
			(res, requirement) =>
				// (a || (b || (c && d)) || e) === (a || b || (c && d) || e)
				requirement.type === 'alternative' ? [...res, ...requirement.requirements] : [...res, requirement],
			[] as Requirement[]
		),
	});
}

export function concatenationRequirementFactory(
	message: string,
	requirements: Requirement[]
): RequirementFactory<ConcatenationRequirement | DependenceRequirement> {
	if (requirements.length < 1) {
		throw new Error('There must be 1 or more requirements');
	}

	if (requirements.length === 1) {
		return dependenceRequirementFactory(message, requirements[0]);
	}

	return (path, value) => ({
		path,
		value,
		message,
		type: 'concatenation',
		requirements: requirements.reduce(
			(res, requirement) =>
				// (a && (b && (c || d)) && e) === (a && b && (c || d) && e)
				requirement.type === 'concatenation' ? [...res, ...requirement.requirements] : [...res, requirement],
			[] as Requirement[]
		),
	});
}

// examples:
// const numberRequirementFactory = singleRequirementFactory('Value should be a number');

// function equalToRequirementFactory(validValue: any): RequirementFactory<SingleRequirement> {
// 	return singleRequirementFactory(`Value should equal to ${validValue}`);
// }

// function optionalRequirementFactory(requirement: Requirement): RequirementFactory<AlternativeRequirement> {
// 	const undefinedFactory = equalToRequirementFactory(undefined);

// 	return (path, value) => alternativeRequirementFactory('', [undefinedFactory(path, value), requirement])(path, value);
// }

// const optionalNumberRequirementFactory: RequirementFactory<AlternativeRequirement> = (path, value) =>
// 	optionalRequirementFactory(numberRequirementFactory(path, value))(path, value);

// function someRequirementFactory(requirement: Requirement): RequirementFactory<DependenceRequirement> {
// 	return dependenceRequirementFactory('Some item should follow the requirement', requirement);
// }
