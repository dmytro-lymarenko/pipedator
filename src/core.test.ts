import { createValidator, pipe, alternative, not, every, valuesByKeys } from './core';

describe('createValidator()', () => {
	it('should create validator', () => {
		const validator = createValidator({
			message: 'Test',
			validate: () => null,
		});

		expect(validator).toBeDefined();
		expect(typeof validator.validate).toBe('function');
	});

	it('should call validate function', () => {
		const validate = jest.fn(() => null);

		const validator = createValidator({
			validate,
			message: 'Test',
		});

		const value = 1;

		const result = validator.validate(value);

		expect(result).toBeNull();
		expect(validate.mock.calls.length).toBe(1);
		expect(validate.mock.calls[0][0]).toBe(value);
	});

	it('should return what validate function returns', () => {
		expect(
			createValidator({
				validate: (value, ctx) => ({
					rootValue: ctx.rootValue,
					errors: [
						{
							value,
							message: ctx.message,
							path: ctx.path,
						},
					],
				}),
				message: 'Test',
			}).validate(1)
		).toMatchSnapshot();
	});
});

describe('pipe()', () => {
	it('should always succeed when validators are not passed down', () => {
		expect(pipe([]).validate(1)).toBeNull();
	});

	it('should invoke passed validator', () => {
		const validate = jest.fn(() => null);

		pipe([createValidator({ validate, message: 'Test' })]).validate(1);

		expect(validate.mock.calls.length).toBe(1);
		expect(validate.mock.calls[0][0]).toBe(1);
	});

	it('should invoke all validators when they all succeed', () => {
		const callbacks = [jest.fn(() => null), jest.fn(() => null), jest.fn(() => null)];

		expect(pipe(callbacks.map(callback => createValidator({ validate: callback, message: 'Test' }))).validate(1)).toBeNull();

		callbacks.forEach(callback => {
			expect(callback.mock.calls.length).toBe(1);
		});
	});

	it('should return the first met error in pipe chain of validators', () => {
		const callbacks = [
			jest.fn(() => null),
			jest.fn(() => (value: any, ctx: any) => ({
				rootValue: ctx.value,
				errors: [{ value, message: ctx.message, path: ctx.path }],
			})),
			jest.fn(() => null),
		];

		expect(
			pipe(callbacks.map((callback, i) => createValidator({ validate: callback, message: `Test #${i}` }))).validate(1)
		).toBeDefined();

		expect(callbacks[0].mock.calls.length).toBe(1);
		expect(callbacks[1].mock.calls.length).toBe(1);
		expect(callbacks[2].mock.calls.length).toBe(0);
	});

	it('should support custom message', () => {
		const callbacks = [
			jest.fn(() => null),
			jest.fn((value: any, ctx: any) => ({
				rootValue: ctx.rootValue,
				errors: [{ value, message: ctx.message, path: ctx.path }],
			})),
		];

		expect(
			pipe(
				callbacks.map(callback => createValidator({ validate: callback, message: 'Test' })),
				'Custom message'
			).validate(1)
		).toMatchSnapshot();
	});
});

describe('alternative()', () => {
	it('should always succeed when validators are not passed down', () => {
		expect(alternative([]).validate(1)).toBeNull();
	});

	it('should invoke passed validator', () => {
		const validate = jest.fn(() => null);

		alternative([createValidator({ validate, message: 'Test' })]).validate(1);

		expect(validate.mock.calls.length).toBe(1);
		expect(validate.mock.calls[0][0]).toBe(1);
	});

	it('should invoke all validators when they all fail', () => {
		const mockCallback = (value: any, ctx: any) => ({
			rootValue: value,
			errors: [
				{
					value,
					message: ctx.message,
					path: ctx.path,
				},
			],
		});

		const callbacks = [jest.fn(mockCallback), jest.fn(mockCallback), jest.fn(mockCallback)];

		expect(
			alternative(callbacks.map(callback => createValidator({ validate: callback, message: 'Test' }))).validate(1)
		).not.toBeNull();

		callbacks.forEach(callback => {
			expect(callback.mock.calls.length).toBe(1);
		});
	});

	it('should succeed when at least one validator succeed', () => {
		const mockCallback = (value: any, ctx: any) => ({
			rootValue: value,
			errors: [
				{
					value,
					message: ctx.message,
					path: ctx.path,
				},
			],
		});

		const callbacks = [jest.fn(mockCallback), jest.fn(() => null), jest.fn(mockCallback)];

		expect(
			alternative(callbacks.map(callback => createValidator({ validate: callback, message: 'Test' }))).validate(1)
		).toBeNull();

		expect(callbacks[0].mock.calls.length).toBe(1);
		expect(callbacks[1].mock.calls.length).toBe(1);
		expect(callbacks[2].mock.calls.length).toBe(0);
	});

	it('should support custom message', () => {
		const mockCallback = (value: any, ctx: any) => ({
			rootValue: ctx.rootValue,
			errors: [
				{
					value,
					message: ctx.message,
					path: ctx.path,
				},
			],
		});

		const callbacks = [jest.fn(mockCallback), jest.fn(mockCallback)];

		expect(
			alternative(
				callbacks.map(callback => createValidator({ validate: callback, message: 'Test' })),
				'Custom message'
			).validate(1)
		).toMatchSnapshot();
	});
});

describe('not()', () => {
	it('should opposite validation', () => {
		expect(
			not(
				createValidator({
					message: 'Always fails',
					validate: (value, ctx) => ({ rootValue: value, errors: [{ value, message: ctx.message, path: ctx.path }] }),
				})
			).validate(1)
		).toBeNull();

		expect(
			not(
				createValidator({
					message: 'Always succeed',
					validate: () => null,
				})
			).validate(1)
		).toMatchSnapshot();
	});

	it('should support custom message', () => {
		expect(
			not(
				createValidator({
					message: 'Always succeed',
					validate: () => null,
				}),
				'Always fails'
			).validate(1)
		).toMatchSnapshot();
	});
});

describe('Complex validator', () => {
	it('should validate', () => {
		const required = createValidator({
			message: 'Value is required',
			validate: (value, ctx) =>
				value === undefined || value === null
					? { rootValue: value, errors: [{ value, message: ctx.message, path: ctx.path }] }
					: null,
		});

		const validator = valuesByKeys(['a', 'b'], every(required), 'Each value should be required');

		console.log(
			validator.validate({
				a: 3,
				b: null,
				c: undefined,
			})
		);
	});
});
