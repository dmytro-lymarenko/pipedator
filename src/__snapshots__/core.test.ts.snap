// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alternative() should support custom message 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Custom message",
      "path": Array [],
      "value": 1,
    },
  ],
  "rootValue": 1,
}
`;

exports[`createValidator() should return what validate function returns 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Test",
      "path": Array [],
      "value": 1,
    },
  ],
  "rootValue": 1,
}
`;

exports[`not() should opposite validation 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Value should not follow the rule: Always succeed",
      "path": Array [],
      "value": 1,
    },
  ],
  "rootValue": 1,
}
`;

exports[`not() should support custom message 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Always fails",
      "path": Array [],
      "value": 1,
    },
  ],
  "rootValue": 1,
}
`;

exports[`pipe() should support custom message 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Custom message",
      "path": Array [],
      "value": 1,
    },
  ],
  "rootValue": 1,
}
`;
